### PR TITLE
Avro Event Format for CloudEvents

### LINKED ISSUE #426: Avro format enconding
I would like to propose an official Avro encoding schema to be supported by CloudEvents.

I've seen a lot of uses cases, in special here in my current project, with Apache Kafka and the many of them were using Avro as the format. Inspired by [this](https://gist.github.com/rbramley/c00bbd95f3f6ecc0cc1f90e5d6d6c90c) Gist (thanks @rbramley), the following schema could be the first draft:

```json
{
  "namespace":"io.cloudevents",
  "type":"record",
  "name":"CloudEvent",
  "version": "1",
  "doc": "CloudEvents Spec for Avro format encoding (the version is just for the schema itself)",
  "fields":[
    {
      "name":"specversion",
      "type":"string"
    },
    {
      "name":"type",
      "type":"string"
    },
    {
      "name":"source",
      "type":"string"
    },
    {
      "name":"id",
      "type":"string",
    },
    {
      "name":"time",
      "type":[
        "null",
        "int"
      ],
      "logicalType": "date"
    },
    {
      "name":"schemaurl",
      "type":[
        "null",
        "string"
      ]
    },
    {
      "name":"contenttype",
      "type":[
        "null",
        "string"
      ]
    },
    {
      "name":"data",
      "type":[
        "null",
        "string"
      ]
    },
    {
      "name":"extensions",
      "type":[
        "null",
        {
          "type":"map",
          "values":[
            "string",
            "int",
            "long",
            "float",
            "double",
            "boolean",
            "bytes"
          ]
        }
      ]
    },
  ]
}
```

- Official Avro spec available [here](https://avro.apache.org/docs/1.8.2/spec.html)

__Extensions__

To define a top-level extension (like below) is not allowed by Avro spec for `avsc`, because we do not have the feature for dynamic fields like JSON. Because of this, you will see the `extension` attribute to the example above.

```json
...
    "comexampleextension1" : "value",
    "comexampleextension2" : {
        "othervalue": 5
    }
...
```

For that, we may employ Avro spec for `ìdl` that is a higher-level language for authoring Avro schemas.

- Official Avro IDL docs available [here](https://avro.apache.org/docs/1.8.2/idl.html)

__Special Types__

What I mean with _special types_, logical types in JSON Schema, are: `uri-reference`. We do not have this type in Avro.

For that, again, we may employ Avro IDL.

### PR DESCRIPTION
**Pull request by fabiojose**
Signed-off-by: Fabio José <fabiojose@gmail.com>

Closes #426

Accidentally I've closed the original PR, but open this new one with the right changes.

### CODE REVIEW

**Comment by NickLarsenNZ** [comment]
Original PR: https://github.com/cloudevents/spec/pull/438

**Comment by duglin** [comment]
@fabiojose can you update the README to point to this doc?

File: avro-format.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+Users of Avro MUST use a message whose binary encoding is identical
+to the one described by the [CloudEvent Avro Schema](./spec.avsc):
+
+```json
+{
+  "namespace":"io.cloudevents",
+  "type":"record",
+  "name":"CloudEvent",
+  "version":"0.3",
```
Comment (duglin) [inline avro-format.md]: 0.4-wip

File: avro-format.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+```
+
+## 3 Examples
+
+The following table shows exemplary mappings:
+
+| CloudEvents     | Type      | Exemplary Avro Value                           |
+| --------------- | --------- | ---------------------------------------------- |
+| type            | string    | `"com.example.someevent"`                      |
+| specversion     | string    | `"0.2"`                                        |
```
Comment (duglin) [inline avro-format.md]: 0.4-wip

File: spec.avsc [hunk 3]
```diff
@@ -0,0 +1,53 @@
+{
+  "namespace":"io.cloudevents",
+  "type":"record",
+  "name":"CloudEvent",
+  "version":"0.2",
```
Comment (duglin) [inline spec.avsc]: 0.4-wip

Comment (fabiojose) [reply spec.avsc]: Done!

File: avro-format.md (hunk 2 above)
Comment (fabiojose) [reply avro-format.md]: Done!

File: avro-format.md (hunk 1 above)
Comment (fabiojose) [reply avro-format.md]: Done!

File: README.md [hunk 4]
```diff
@@ -38,6 +38,7 @@ The following documents are available:
 | NATS Transport Binding        | [v0.3](https://github.com/cloudevents/spec/blob/v0.3/nats-transport-binding.md) | [master](https://github.com/cloudevents/spec/blob/master/nats-transport-binding.md) |
 | Protocol Buffers Event Format | [v0.3](https://github.com/cloudevents/spec/blob/v0.3/protobuf-format.md)        | [master](https://github.com/cloudevents/spec/blob/master/protobuf-format.md)        |
 | Web hook                      |      [v0.3](https://github.com/cloudevents/spec/blob/v0.3/http-webhook.md)      |      [master](https://github.com/cloudevents/spec/blob/master/http-webhook.md)      |
+| AVRO Event Format             | - | [master](./avro-format.md) |
```
Comment (duglin) [inline README.md]: can you move this so the list if alphabetized?

**Comment by duglin** [comment]
We tentatively approved this on today's call, but people asked for one more week to make sure. Please look it over so we can resolve it next week

File: README.md (hunk 4 above)
Comment (duglin) [reply README.md]: @fabiojose can you also add this to the Makefile so it gets verified?

**Comment by duglin** [comment]
@JemDay I believe you asked for one more week,last week.  You ok with this one?

File: README.md (hunk 4 above)
Comment (fabiojose) [reply README.md]: @duglin  done!

File: avro-format.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+map-of-maps in a recursive way, i.e., every `map` declaration REQUIRE the
+`values` with primitives types or a declared type.
+
+### 2.3 Optional Attributes
+
+CloudEvents Spec defines optional attributes. The Avro format defines that
+every optional field MUST use the `null` type and the actual type through
+the [union][avro-unions].
+
+Declaring a type which may be either a `null` or `string`, i.e., an optional
```
Comment (duglin) [inline avro-format.md]: s/may/might/

File: avro-format.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+attributes they define.
+
+### 2.2 Mapping Any-typed Attributes
+
+`Any`-typed CloudEvents values can either hold a `String`, or a `Binary` value,
+or a `Map`, or any of all other types. Avro type system SHOULD NOT satisfy the
+map-of-maps in a recursive way, i.e., every `map` declaration REQUIRE the
+`values` with primitives types or a declared type.
+
+### 2.3 Optional Attributes
```
Comment (duglin) [inline avro-format.md]: s/Optional/OPTIONAL/

File: avro-format.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+CloudEvents Spec defines optional attributes. The Avro format defines that
+every optional field MUST use the `null` type and the actual type through
```
Comment (duglin) [inline avro-format.md]: s/every optional field/these fields/

File: avro-format.md (hunk 5 above)
Comment (duglin) [reply avro-format.md]: is the above a complete sentence? We need to do something about "optional", the RFC checker isn't happy. I would suggest we either remove the "i.e" part, if that still makes sense

Comment (fabiojose) [reply avro-format.md]: Did not make sense! Removed!

**Comment by clemensv** [comment]
I don't think the definition of a CloudEvent as a "record" satisfies our need for having extensibility, because record requires all fields to be known for encoding.

**Comment by JemDay** [comment]
I think the prior comment from @clemensv may be touching slightly on what i'd been thinking about given proposal was very structured (which in a way i like) vs the approach taken with proto which way much more minimalistic. The proto schema is more aligned with the example @clemensv cited.

As it stands with the current schema proposal a mechanism to carry 'extensions' is required.

**Comment by fabiojose** [comment]
Very clever @clemensv 

I tested it and got the following schema:

```json
{
  "namespace":"io.cloudevents",
  "type":"record",
  "name":"CloudEvent",
  "version":"0.4-wip",
  "doc":"Avro Event Format fo CloudEvents",
  "fields":[
    {
      "name":"metadata",
      "type":{
        "type":"map",
        "values":[
          "null",
          "int",
          "float",
          "string",
          "boolean",
          "long",
          "CloudEvent"
        ]
      }
    }
  ]
}
```

It looks more flexible than the first version.

**Comment by evankanderson** [comment]
We don't have `float` or `boolean` types defined at the moment.

We may also need to support `binary` types.

**Comment by fabiojose** [comment]
Good points @evankanderson!

Done!

**Comment by duglin** [comment]
@clemensv @evankanderson does this look ok now?

**Comment by JemDay** [comment]
Disclaimer : Not an avro expert..

For consistency reasons with protobuf...

- Could we change the name of the definition file to _cloudevent.avsc_
- Does avro allow you to follow the same model as used in cloudevent.proto

Otherwise +1

**Comment by fabiojose** [comment]
Hi @JemDay 

"Could we change the name of the definition file to cloudevent.avsc": Sure!
- I had used the `spec.avsc` to follow the same naming convention used by `spec.json`.

"Does avro allow you to follow the same model as used in cloudevent.proto"
- unfortunately, not

**Comment by JemDay** [comment]
Thanks ... thinking a bit more, maybe aligning around _spec.<format>_ is better

Still +1

**Comment by fabiojose** [comment]
Good point! I will change the file name.

**Comment by fabiojose** [comment]
I think that `cloudevent.avsc` is better.

**Comment by duglin** [comment]
at this point I think consistency with the others is more important tho

**Comment by duglin** [comment]
@fabiojose is this one ready to go? it was approved to but I don't want to merge if you still have a pending change

**Comment by fabiojose** [comment]
Done, no pending changes!

The file will be `spec.avsc`, just like before.

**Comment by catkins-miso** [comment]
It seems there would have been a performance impact removing the first-class context attributes in favor of the attribute map approach. I understand the point of adding the attribute map for extensions, but I think the schema-on-read approach could have been isolated without impacting other use cases. In particular, if I'm listening to a firehose topic (events of many types)--or implementing an intermediary--I need to quickly evaluate the type context attribute. Not to mention Avro has a pretty mechanics for [forward and backwards compatibility](https://avro.apache.org/docs/current/spec.html#Schema+Resolution) which I _think_ could have addressed concerns about introducing new context attributes.

**Comment by evankanderson** [comment]
I think the attributes may have been simplified in the interim for v1 (when this was written, they were a json-style `string` -> `value or object` mapping, rather than a `string` -> `simple type` mapping.

Unfortunately, I think the v1.0 spec ship has sailed; I'm not sure what to suggest here.

**Comment by catkins-miso** [comment]
Thanks for the response @evankanderson. I did a little preliminary benchmarking (Apache Avro 1.9 dotnetcore), and I am seeing no meaningful difference in the binary serde performance from the attributes and having distinct fields.

**Pull request merged**